sql: new schema changer does not properly use phases for drops

fqazi opened the pull request:
Fixes: #66613

Previously, when drop operations were added into the new schema
changer we added all operations into the statement phase. This
was inadequate because if an explicit transaction dropped
a table and then most of the work would be applied before
the transaction was committed potentially influencing
other transactions. To address this, this patch splits the drop
into between statement and pre-commit phases properly. In
the pre-commit phase, a synthetic descriptor marks the
object as dropped, so that only the transaction sees the
drop. The pre-commit phase which executes before committing
then marks the descriptor as dropped.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68616)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
Cool stuff 

> Reviewed 6 of 21 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-:-MgghjdZB95Mf164JxtH:b95tn32)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[pkg/sql/schema_change_plan_node.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MgggUcv7vbGQ0g-9ahi:-MgggUcv7vbGQ0g-9ahj:b-v9mgxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/schema_change_plan_node.go#L36)):*
> ```Go
> 	// TODO(ajwerner): Call featureflag.CheckEnabled appropriately.
> 	mode := p.extendedEvalCtx.SchemaChangerState.mode
> 	// When new schema changer is on we wil not support it for explicit
> ```

`s/wil/will/`
___
*[pkg/sql/catalog/descs/collection.go, line 346 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mgggad9C-puufAil3Zs:-Mgggad9C-puufAil3Zt:b8fgm0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/catalog/descs/collection.go#L346)):*
<details><summary><i alt="summary">Quoted 11 lines of code…</i></summary>

> ```
> // AddSyntheticDescriptorForTxn replaces a descriptor with a synthetic
> // one temporarily for a given transaction. This synthetic descriptor
> // will be immutable.
> func (tc *Collection) AddSyntheticDescriptorForTxn(desc catalog.Descriptor) {
> 	tc.synthetic.add(desc)
> }
> // RemoveSyntheticDescriptorForTxn removes a synthetic descriptor
> // override that temporarily exists for a given transaction.
> func (tc *Collection) RemoveSyntheticDescriptorForTxn(id descpb.ID) {
> 	tc.synthetic.remove(id)
> }
> ```
</details>

I find `ForTxn` confusing. 
___
*[pkg/sql/catalog/descs/synthetic_descriptors.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mgggjk-874iRHRdVpZo:-Mgggjk-874iRHRdVpZp:b9mzmnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/catalog/descs/synthetic_descriptors.go#L44)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 		if mut, ok := desc.(catalog.MutableDescriptor); ok {
> 			desc = mut.ImmutableCopy()
> 		}
> 		sd.descs.Upsert(desc)
> ```
</details>

should this just call `add` now?
___
*[pkg/sql/logictest/testdata/logic_test/new_schema_changer, line 654 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mgggs2X0QknjHHAbwJP:-Mgggs2X0QknjHHAbwJQ:b-x3snzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/logictest/testdata/logic_test/new_schema_changer#L654)):*
> ```
> 
> ROLLBACK
> ```
Once you've rolled back, you aren't in the transaction any more. 
___
*[pkg/sql/schemachanger/scop/mutation.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mggh9Pt9i1B106VznnB:-Mggh9Pt9i1B106VznnC:b-p2r5jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/schemachanger/scop/mutation.go#L73)):*
> ```Go
> type MarkDescriptorAsDroppedInTxn struct {
> 	mutationOp
> 	TableID descpb.ID
> ```

`DescID` as opposed to `TableID`?
___
*[pkg/sql/schemachanger/scpb/scpb.proto, line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MgghD9tBVunJiS8RDaN:-MgghD9tBVunJiS8RDaO:b-v8fv9o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/schemachanger/scpb/scpb.proto#L58)):*
> ```protobuf
>   BACKFILLED = 4;
>   VALIDATED = 5;
>   TXN_DROPPED = 6;
> ```

What about `IN_MEMORY_DROPPED` or `SYNTHETICALLY_DROPPED`?
___
*[pkg/sql/schemachanger/scplan/rules.go, line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MgghQbg7V3Eyt6U25rf:-MgghQbg7V3Eyt6U25rg:bqyy3p6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/schemachanger/scplan/rules.go#L241)):*
<details><summary><i alt="summary">Quoted 7 lines of code…</i></summary>

> ```
> 				{
> 					predicate: func(this *scpb.Database, flags Params) bool {
> 						return flags.ExecutionPhase == StatementPhase &&
> 							!flags.CreatedDescriptorIDs.Contains(this.DatabaseID)
> 					},
> 				},
> 				{
> ```
</details>

My sense is that we do want to do this in the statement phase, no?
___
*[pkg/sql/schemachanger/scplan/rules.go, line 257 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MgghVVSEzkjqOFFCBPD:-MgghVVSEzkjqOFFCBPE:b-x0yw61) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/schemachanger/scplan/rules.go#L257)):*
> ```Go
> 				{
> 					predicate: func(this *scpb.Database, flags Params) bool {
> 						return flags.ExecutionPhase == PreCommitPhase &&
> ```

This one is interesting, I think we don't want to do it before `PreCommit` which means this would say `StatementPhase` but we can do this in `PreCommit` if there's nothing preventing it. 


<!-- Sent from Reviewable.io -->

fqazi committed:
sql: new schema changer does not properly use phases for drops

Fixes: #66613

Previously, when drop operations were added into the new schema
changer we added all operations into the statement phase. This
was inadequate because if an explicit transaction dropped
a table and then most of the work would be applied before
the transaction was committed potentially influencing
other transactions. To address this, this patch splits the drop
into between statement and pre-commit phases properly. In
the pre-commit phase, a synthetic descriptor marks the
object as dropped, so that only the transaction sees the
drop. The pre-commit phase which executes before committing
then marks the descriptor as dropped.

Release note: None

fqazi committed:
sql: new schema changer does not correctly sort operations

Previously, the new schema changer had bugs in the sort
order which would lead to dependency edges not sorting
things correctly. Additionally, revertability should also
be inherited from these edges. This was inadequate because
operations did note execute in the correct order. To address
this, this patch fixes various sort related bugs and adds
support for inheriting revertability based on dependencies.

Release note: None

ajwerner reviewed (commented):
> Reviewed 5 of 22 files at r3, 1 of 30 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/catalog/descs/collection.go, line 363 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmT0PYxEApelqRpR9Eb:-MmT0PYxEApelqRpR9Ec:bgjx4ub) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ebbdb2ca5abe5f5fb6573ff86a63c92d37e67e/pkg/sql/catalog/descs/collection.go#L363)):*
> ```Go
> // RemoveSyntheticDescriptorForTxn removes a synthetic descriptor
> // override that temporarily exists for a given transaction.
> func (tc *Collection) RemoveSyntheticDescriptorForTxn(id descpb.ID) {
> ```

why `ForTxn`?
___
*[pkg/sql/logictest/testdata/logic_test/new_schema_changer, line 654 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mgggs2X0QknjHHAbwJP:-MmT0ZIm-aEVhYis5GWr:bt4xwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/logictest/testdata/logic_test/new_schema_changer#L654)):*
> statement error current transaction is aborted, commands ignored until end of transaction block
> SELECT * from db1.sc1.sq1
>   
>
> statement error current transaction is aborted, commands ignored until end of transaction block
> SELECT 'a'::db1.sc1.typ::string
>   
  
are these testing anything interesting?
___
*[pkg/sql/schemachanger/scop/phase.go, line 34 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmT0sIC1M2zXLtz9hZl:-MmT0sIC1M2zXLtz9hZm:b-rsh47q) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ebbdb2ca5abe5f5fb6573ff86a63c92d37e67e/pkg/sql/schemachanger/scop/phase.go#L34)):*
> ```Go
> 	// PostCommitPhase refers to execution of ops occurring after the user
> 	// transaction has committed (i.e., in the async schema change job).
> 	// Note: Planning rules cannot ever be in this phase, since all those operations
> ```

what's a planning rule?
___
*[pkg/sql/schemachanger/scplan/plan.go, line 153 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmT-u7G-k8isrhT7tjx:-MmT-u7G-k8isrhT7tjy:b-7pjmzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46d953cdbe3df1fd5a9cac73759c223fea15528/pkg/sql/schemachanger/scplan/plan.go#L153)):*
> ```Go
> 						opEdge, ok := g.GetOpEdgeTo(de.To())
> 						if ok {
> 							revertibleDeps = revertibleDeps && opEdge.Revertible()
> ```

Should this be `revertibleDeps = revertibleDeps || opEdge.Revertible()`?
___
*[pkg/sql/schemachanger/scplan/plan.go, line 157 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmT-yg0EQlp9O5IdHEF:-MmT-yg0EQlp9O5IdHEG:bsu8qf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46d953cdbe3df1fd5a9cac73759c223fea15528/pkg/sql/schemachanger/scplan/plan.go#L157)):*
> ```Go
> 						return nil
> 					})
> 					isOpEdgeRevertible := revertibleDeps && e.Revertible()
> ```

should this be `||`?
___
*[pkg/sql/schemachanger/scplan/plan.go, line 268 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmT0ErZAlyeQU6frzmf:-MmT0ErZAlyeQU6frzmg:bm22rbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46d953cdbe3df1fd5a9cac73759c223fea15528/pkg/sql/schemachanger/scplan/plan.go#L268)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 	// Unfortunately, we are forced to do an inefficient
> 	// bubble sort, since with unrelated dependencies will
> 	// be equal to each other. But, still have a relative order
> 	// across the entire set of nodes.
> ```
</details>


increasingly I'm thinking we need to revisit both this scheme and its implementation


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schema_change_plan_node.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MgggUcv7vbGQ0g-9ahi:-MgkKybi-55loEkLO7Tr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/schema_change_plan_node.go#L36)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`s/wil/will/`
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descs/collection.go, line 346 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mgggad9C-puufAil3Zs:-MgkKzFJ8SrFpUAQGEMJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/catalog/descs/collection.go#L346)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> // AddSyntheticDescriptorForTxn replaces a descriptor with a synthetic
> // one temporarily for a given transaction. This synthetic descriptor
> // will be immutable.
> func (tc *Collection) AddSyntheticDescriptorForTxn(desc catalog.Descriptor) {
> 	tc.synthetic.add(desc)
> }
> // RemoveSyntheticDescriptorForTxn removes a synthetic descriptor
> // override that temporarily exists for a given transaction.
> func (tc *Collection) RemoveSyntheticDescriptorForTxn(id descpb.ID) {
> 	tc.synthetic.remove(id)
> }
> ```
I find `ForTxn` confusing. 
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/new_schema_changer, line 654 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mgggs2X0QknjHHAbwJP:-MgkL1lpAe27ERD34hgp:b-sox6gz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/logictest/testdata/logic_test/new_schema_changer#L654)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> statement error current transaction is aborted, commands ignored until end of transaction block
> SELECT * from db1.sc1.sq1
>   
>
> statement error current transaction is aborted, commands ignored until end of transaction block
> SELECT 'a'::db1.sc1.typ::string
>   
  
are these testing anything interesting?

</blockquote></details>

Nope this was left over cleaned it up.
___
*[pkg/sql/schemachanger/scop/mutation.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mggh9Pt9i1B106VznnB:-MgkL9dJ3q2EwiD-20Dv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/schemachanger/scop/mutation.go#L73)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`DescID` as opposed to `TableID`?
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scop/phase.go, line 34 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmT0sIC1M2zXLtz9hZl:-MmTnO9d2RyB8Eg9-qko:b-1gtqtk) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ebbdb2ca5abe5f5fb6573ff86a63c92d37e67e/pkg/sql/schemachanger/scop/phase.go#L34)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

what's a planning rule?
</blockquote></details>

Removed this it was referring to predicates never getting POST COMMIT as a minimum
___
*[pkg/sql/schemachanger/scplan/plan.go, line 153 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmT-u7G-k8isrhT7tjx:-MmTnUGb-vT0ryQlzqou:b-60nse4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46d953cdbe3df1fd5a9cac73759c223fea15528/pkg/sql/schemachanger/scplan/plan.go#L153)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Should this be `revertibleDeps = revertibleDeps || opEdge.Revertible()`?
</blockquote></details>

Discussed this over Slack. Side effect of false having a special meaning.
___
*[pkg/sql/schemachanger/scplan/plan.go, line 268 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmT0ErZAlyeQU6frzmf:-MmTndR69Iu4Q985Kigg:b-1vy6my) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46d953cdbe3df1fd5a9cac73759c223fea15528/pkg/sql/schemachanger/scplan/plan.go#L268)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// Unfortunately, we are forced to do an inefficient
> 	// bubble sort, since with unrelated dependencies will
> 	// be equal to each other. But, still have a relative order
> 	// across the entire set of nodes.
> ```

increasingly I'm thinking we need to revisit both this scheme and its implementation
</blockquote></details>

Yes, I think we can remove the sort if we fix some of the rule evaluation stuff. Pretty much always guarantee dependency order is enforced when emitting ops.
___
*[pkg/sql/schemachanger/scplan/rules.go, line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MgghQbg7V3Eyt6U25rf:-MgkLAym9d11gV0Pvc3l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae4c61fa747a5041a7badcfeea9b235d5bbc52c6/pkg/sql/schemachanger/scplan/rules.go#L241)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 				{
> 					predicate: func(this *scpb.Database, flags Params) bool {
> 						return flags.ExecutionPhase == StatementPhase &&
> 							!flags.CreatedDescriptorIDs.Contains(this.DatabaseID)
> 					},
> 				},
> 				{
> ```
My sense is that we do want to do this in the statement phase, no?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Reviewed 1 of 30 files at r5, 14 of 29 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/sql/lex/BUILD.bazel, line 8 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY3WVXCxjwYnR_Dv6n:-MmY3WVXCxjwYnR_Dv6o:b3y4t66) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/lex/BUILD.bazel#L8)):*
> ```
>         "keywords.go",
>         "tokens.go",
> ```
I think this is detritus
___
*[pkg/sql/schemachanger/scexec/mutation_desc_getter.go, line 113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY3deX6NzIKiQotliZ:-MmY3deX6NzIKiQotli_:b-l1j4si) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scexec/mutation_desc_getter.go#L113)):*
> ```Go
> }
> 
> func (m *mutationDescGetter) MarkDescriptorAsDroppedInTxn(ctx context.Context, id descpb.ID) error {
> ```

nit: how do you feel about using the term `Synthetic` here? 
___
*[pkg/sql/schemachanger/scop/phase.go, line 34 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmT0sIC1M2zXLtz9hZl:-MmY43og7rPkl4n5hS8h:b-th5tha) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ebbdb2ca5abe5f5fb6573ff86a63c92d37e67e/pkg/sql/schemachanger/scop/phase.go#L34)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Removed this it was referring to predicates never getting POST COMMIT as a minimum
</blockquote></details>

```
	// Note: Planning rules cannot ever be in this phase, since all those operations
	// should be executed in pre-commit.
```

I'm still not clear on what this means. Planning rule is not a term I have a clear definition of. 
___
*[pkg/sql/schemachanger/scpb/scpb.proto, line 59 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4Kc82T8Cd7efrnxy:-MmY4Kc93K031gm8ECqM:b-7tv5hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scpb/scpb.proto#L59)):*
> ```
>   TXN_DROPPED = 6;
>   PUBLIC = 7;
> ```
soon we're going to need to give up on re-ordering these
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 50 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4sV09rgSkjtAxXys:-MmY4sV09rgSkjtAxXyt:b-okhe26) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L50)):*
> ```Go
> 		other, otherTarget, otherNode    = targetNodeVars("other")
> 	)
> 	// TODO(ajwerner): These rules are suspect.
> ```

fix up this comment now that it makes more sense
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 110 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4q383iOKBLanxKVy:-MmY4q383iOKBLanxKVz:b-okhe26) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L110)):*
> ```Go
> 	ownedBy, ownedByTarget, ownedByNode := targetNodeVars("owned-by")
> 	id := rel.Var("id")
> 	// TODO(ajwerner): What is this about? It seems, if anything, just wrong.
> ```

fix up this comment now that it makes more sense
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 308 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4ot91-JAT4euwgkl:-MmY4ot91-JAT4euwgkm:b-okhe26) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L308)):*
> ```Go
> 
> 	// TODO(ajwerner): What does this even mean? The table starts in
> 	// public.
> ```

fix up this comment now that it makes more sense
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 331 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4lA15mX-d3LTtP-c:-MmY4lA15mX-d3LTtP-d:b-72ozye) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L331)):*
> ```
> 	// TODO(ajwerner): What does this even mean? The table starts in
> 	// public.
> ```
fix up this comment now that it makes more sense
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 354 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4tby6ZP9IJCm0lJq:-MmY4tby6ZP9IJCm0lJr:b-okhe26) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L354)):*
> ```Go
> func init() {
> 	// These are more weird rules where the public drop table
> 	// depends on something else.
> ```

fix up this comment now that it makes more sense
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 378 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4zzTD1lRwsBQ6tif:-MmY4zzTD1lRwsBQ6tig:bpaf1vt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L378)):*
> ```
> 	// Require that a table has entered an absent state
> 	// if its being dropped at the same time. This ensures
> 	// that we transitioned away from txnDropped state.
> ```
When we add a new phase for the drop table to separate the draining names, is it reasonable to say that the fk status we care about is going to be `DROPPED` as opposed to `ABSENT`?
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_table.go, line 39 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY5b9f471MyJcv6MaL:-MmY5b9f471MyJcv6MaM:b-dkp3ot) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/opgen/opgen_table.go#L39)):*
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>

> ```
> 			emit(func(this *scpb.Table) scop.Op {
> 				return &scop.DrainDescriptorName{
> 					TableID: this.TableID,
> 				}
> 			}),
> 			emit(func(this *scpb.Table) scop.Op {
> 				return &scop.CreateGcJobForDescriptor{
> 					DescID: this.TableID,
> 				}
> 			}),
> ```
</details>


Looking back at these rules, we can't drain the name until after we've dropped the table. We need to add another phase here. Something like `Status_DROPPED` followed by `Status_ABSENT`. I suspect this applies to other drop rules. At least, we need to do that if we want parity with the existing draining names behavior. Even if we don't, I don't think we should create the GC job until we know that no other node views the table as public.  


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 50 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4sV09rgSkjtAxXys:-MmYjwPNARTEaK0z4MiI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L50)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

fix up this comment now that it makes more sense
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 110 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4q383iOKBLanxKVy:-MmYjx-s9LvxZZScPmh3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L110)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

fix up this comment now that it makes more sense
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 308 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4ot91-JAT4euwgkl:-MmYjxck3hd-5l6Hhp0U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L308)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

fix up this comment now that it makes more sense
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 331 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4lA15mX-d3LTtP-c:-MmYjy6g3iYKK7AC0PXV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L331)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// TODO(ajwerner): What does this even mean? The table starts in
> 	// public.
> ```
fix up this comment now that it makes more sense
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 354 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4tby6ZP9IJCm0lJq:-MmYjyiXAXebSU4I_Iay:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L354)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

fix up this comment now that it makes more sense
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 378 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4zzTD1lRwsBQ6tif:-MmYjzdO5psGPkj1Vegx:b-wchvgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L378)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// Require that a table has entered an absent state
> 	// if its being dropped at the same time. This ensures
> 	// that we transitioned away from txnDropped state.
> ```
When we add a new phase for the drop table to separate the draining names, is it reasonable to say that the fk status we care about is going to be `DROPPED` as opposed to `ABSENT`?
</blockquote></details>

Added a new DROPPED state before absent, and adjust + simplified rules.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/sql/schemachanger/scplan/testdata/alter_table, line 47 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mms6EyE6MF4IhfO68T-:-Mms6EyE6MF4IhfO68T0:b-gmyy8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/835713327bfe826ff720985d025dc8fc96103afa/pkg/sql/schemachanger/scplan/testdata/alter_table#L47)):*
<details><summary><i alt="summary">Quoted 17 lines of code…</i></summary>

> ```
> Stage 2 (non-revertible)
>   *scop.MakeAddedColumnDeleteOnly
>     Column:
>       ID: 2
>       Name: j
>       Nullable: true
>       Type:
>         family: IntFamily
>         oid: 20
>         width: 64
>     FamilyName: primary
>     TableID: 52
> Stage 3 (non-revertible)
>   *scop.MakeAddedColumnDeleteAndWriteOnly
>     ColumnID: 2
>     TableID: 52
> Stage 4
> ```
</details>

There's something not right here, namely, I think we need to make sure that once a stage becomes non-revertible, all subsequent stages are non-revertible. I'm worried the plan as written is somehow impossible with that restriction.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Reviewed 3 of 34 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/sql/schemachanger/scgraph/graph.go, line 218 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmyEJoRC2hPiSMgACJ6:-MmyEJoRC2hPiSMgACJ7:bcvawjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/976a58e9857cbbbc33ee23dd172269820202ac0a/pkg/sql/schemachanger/scgraph/graph.go#L218)):*
> ```Go
> 	// Revertibility should also be inherited from the source edge.
> 	if sourceOpEdge, sourceOk := g.GetOpEdgeTo(de.from); sourceOk && !sourceOpEdge.revertible {
> 		if targetOpEdge, targetOk := g.GetOpEdgeTo(de.to); targetOk && targetOpEdge.revertible {
> ```

Is this needed given the change in opgen?
___
*[pkg/sql/schemachanger/scplan/opgen/op_gen.go, line 95 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmyF-9_7j1m4lwhcmjk:-MmyF-9_7j1m4lwhcmjl:b-8ndzee) ([raw file](https://github.com/cockroachdb/cockroach/blob/976a58e9857cbbbc33ee23dd172269820202ac0a/pkg/sql/schemachanger/scplan/opgen/op_gen.go#L95)):*
<details><summary><i alt="summary">Quoted 7 lines of code…</i></summary>

> ```
> 			// If any previous transition was non-revertible,
> 			// we should inherit that here.
> 			if previousNode, ok := g.GetNode(op.n.Target, op.from); ok {
> 				if edge, ok := g.GetOpEdgeTo(previousNode); ok {
> 					op.revertible = edge.Revertible() && op.revertible
> 				}
> 			}
> ```
</details>

rather than doing this here and looking at the graph itself, how would you feel about doing it above before you add it to the set of edgesToAdd like:

```go
--- a/pkg/sql/schemachanger/scplan/opgen/op_gen.go
+++ b/pkg/sql/schemachanger/scplan/opgen/op_gen.go
@@ -75,7 +75,10 @@ func (r *registry) buildGraph(currentPhase scop.Phase, initial scpb.State) (*scg
                                                op.from = op.to
                                                op.ops = func(element scpb.Element) []scop.Op { return nil }
                                        }
-
+                                       // Revertibility is infectious.
+                                       if len(edgesToAdd) > 0 && !edgesToAdd[len(edgesToAdd)-1].revertible {
+                                               op.revertible = false
+                                       }
                                        edgesToAdd = append(edgesToAdd, toAdd{
                                                transition: op,
                                                n:          n,
```


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Reviewed 1 of 34 files at r11, 9 of 40 files at r13, 26 of 31 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 50 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4sV09rgSkjtAxXys:-Mn0vvjp2lGkR5V4EihK:bulygo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L50)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Is it?
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 258 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn0y9VIAIfeToBkrHLj:-Mn0y9VIAIfeToBkrHLk:benfllo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/deprules/rules.go#L258)):*
> ```Go
> 	relationNeedsDepToBeRemoved := func(ruleName string, depTypes []interface{}, depDescIDMatch rel.Attr) {
> 		// Before any parts of a relation can be dropped, the relation
> 		// should exit the synthetic drop state.
> ```

This comment doesn't align with my expectations. Or, rather, is there no drop state for these descriptors? It feels like maybe we'd want to drop the table and then remove the references? Honestly it feels like we'd want to do that at the same time. I'm not totally sure how to represent it. Suffice it to say, we're not done refining all of this stuff or explaining it. I'm not going to block this review but we have a lot more work to do to explain the flows of these schema changes in commentary from principles which we can reusably apply. 


I guess what I'm getting at is that I'd think we'd go from synthetically dropped to dropped for the relation itself and that we'd make the removal of the parts of the relation depend on the relation being in the dropped state. Then we'd make the relation being absent depend on the parts of the relation being absent (like we have here). That would, I'd think, constrain the removal of the elements to the transaction that moves the relation to dropped. 
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_column.go, line 25 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn0zqWc55zTLP8T2mMJ:-Mn0zqWc55zTLP8T2mMK:b-l26a9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/opgen/opgen_column.go#L25)):*
> ```Go
> 		to(scpb.Status_DELETE_ONLY,
> 			minPhase(scop.PreCommitPhase),
> 			revertible(false),
> ```

I don't think this is right. This is the add direction. This is totally revertible. All of these steps except public are revertible. 
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_column.go, line 63 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1-2BY5jlJcxTWxJtf:-Mn1-2BY5jlJcxTWxJtg:b-o18iaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/opgen/opgen_column.go#L63)):*
> ```Go
> 			})),
> 		to(scpb.Status_DELETE_AND_WRITE_ONLY,
> 			// TODO(ajwerner): This should probably be marked as non-revertible.
> ```

this should be marked as non-revertible while you're here. 
___
*[pkg/sql/schemachanger/scplan/testdata/alter_table, line 31 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1-eIS13xvSxd_jQhC:-Mn1-eIS13xvSxd_jQhD:b-e8ogcd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/testdata/alter_table#L31)):*
> ```
> Stage 2 (non-revertible)
>   *scop.MakeAddedColumnDeleteOnly
> ```
this isn't right, for reasons noted in the `opgen_column.go`.

On a more deep level, if we ever are force to have a non-revertible phase followed by a revertible phase, we have a bug. The `Plan` function should return an assertion failure error rather than letting this happen. 
___
*[pkg/sql/schemachanger/scplan/testdata/drop_database, line 53 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn10-uXBlAl5LPic2rc:-Mn10-uXBlAl5LPic2rd:b8nng9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/testdata/drop_database#L53)):*
> ```
> DROP DATABASE db1 CASCADE
> ----
> Stage 0 (non-revertible)
> ```

why is this non-revertible? 
___
*[pkg/sql/schemachanger/scplan/testdata/drop_database, line 86 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1039i6t8H4su4uBT0:-Mn1039i6t8H4su4uBT1:b8rpx0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/testdata/drop_database#L86)):*
<details><summary><i alt="summary">Quoted 7 lines of code…</i></summary>

> ```
> 
>   *scop.RemoveTypeBackRef
>     DescID: 64
>     TypeID: 62
>   *scop.RemoveTypeBackRef
>     DescID: 64
>     TypeID: 63
> ```
</details>

I don't think this should happen before the relevant descriptor is actually marked as dropped. 


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 50 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-MmY4sV09rgSkjtAxXys:-Mn123R97SxQ5dHjL-K9:b-lmd835) ([raw file](https://github.com/cockroachdb/cockroach/blob/f663909d8a29fc63bbf9e306a01f8c0092d1c3f1/pkg/sql/schemachanger/scplan/deprules/rules.go#L50)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Is it?
</blockquote></details>

I missed this spot. Clarified what these rules entail.
___
*[pkg/sql/schemachanger/scplan/deprules/rules.go, line 258 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn0y9VIAIfeToBkrHLj:-Mn1DlvT470M09z9zSUx:bbzfjzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/deprules/rules.go#L258)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This comment doesn't align with my expectations. Or, rather, is there no drop state for these descriptors? It feels like maybe we'd want to drop the table and then remove the references? Honestly it feels like we'd want to do that at the same time. I'm not totally sure how to represent it. Suffice it to say, we're not done refining all of this stuff or explaining it. I'm not going to block this review but we have a lot more work to do to explain the flows of these schema changes in commentary from principles which we can reusably apply. 


I guess what I'm getting at is that I'd think we'd go from synthetically dropped to dropped for the relation itself and that we'd make the removal of the parts of the relation depend on the relation being in the dropped state. Then we'd make the relation being absent depend on the parts of the relation being absent (like we have here). That would, I'd think, constrain the removal of the elements to the transaction that moves the relation to dropped. 
</blockquote></details>

I tweaked the rules and states so that dependencies now do their work in the dropped state, before hitting absent. That will also guarantee things always end up in the same stages if we have funny scenarios. Let me know if that looks good.
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_column.go, line 25 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn0zqWc55zTLP8T2mMJ:-Mn13GgXCKqdRTWMqtZW:b-q2ndji) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/opgen/opgen_column.go#L25)):*
> depend on the parts of the relation being absent (like we have here). That would
Sigh I mixed it up last night when I was updating.
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_column.go, line 63 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1-2BY5jlJcxTWxJtf:-Mn13LrgE-5TXGpyh_la:b-2kygz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/opgen/opgen_column.go#L63)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this should be marked as non-revertible while you're here. 
</blockquote></details>

This was the correct spot, the to here is also wrong.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Reviewed 3 of 37 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/sql/schemachanger/scplan/opgen/op_gen.go, line 89 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1FIQC9VLW-3So9y8N:-Mn1FIQC9VLW-3So9y8O:b-v7mk7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/op_gen.go#L89)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> 			// Revertibility is infectious and will impact
> 			// all edges that follow.
> 			if len(edgesToAdd) > 0 && !edgesToAdd[len(edgesToAdd)-1].revertible {
> 				op.revertible = false
> 			}
> ```
</details>

I thought about this some more and I no longer think it works. We don't currently add nodes that correspond to already completed stages. I think we should actually change this logic in a different place:

```go
--- a/pkg/sql/schemachanger/scplan/opgen/target.go
+++ b/pkg/sql/schemachanger/scplan/opgen/target.go
@@ -54,7 +54,7 @@ func populateSpecs(status scpb.Status, specs []transitionSpec) {
                        continue
                }
                specs[i].from = specs[i-1].to
-
+               specs[i].revertible = specs[i].revertible && specs[i-1].revertible
                prevMinPhase := specs[i-1].minPhase
                switch curMinPhase := specs[i].minPhase; curMinPhase {
                case 0:
```
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_default_expression.go, line 39 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1ETLtEZRXIjw8XZZT:-Mn1ETLtEZRXIjw8XZZU:b199s50) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_default_expression.go#L39)):*
<details><summary><i alt="summary">Quoted 17 lines of code…</i></summary>

> ```
> 		to(scpb.Status_DROPPED,
> 			minPhase(scop.PreCommitPhase),
> 			revertible(false),
> 			emit(func(this *scpb.DefaultExpression) scop.Op {
> 				return &scop.RemoveColumnDefaultExpression{
> 					TableID:  this.TableID,
> 					ColumnID: this.ColumnID,
> 				}
> 			}),
> 			emit(func(this *scpb.DefaultExpression) scop.Op {
> 				return &scop.UpdateRelationDeps{
> 					TableID: this.TableID,
> 				}
> 			})),
> 		to(scpb.Status_ABSENT,
> 			minPhase(scop.PreCommitPhase),
> 			revertible(false)),
> ```
</details>

I'm not sure I understand the addition of this new phase here. I don't know what `Dropped` means here. 
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_inbound_foreign_key.go, line 34 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1EbYN5Uwmm7cvGlwp:-Mn1EbYN5Uwmm7cvGlwq:b-938lea) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_inbound_foreign_key.go#L34)):*
> ```Go
> 				}
> 			})),
> 		to(scpb.Status_ABSENT,
> ```

ditto comment from default expr
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_outbound_foreign_key.go, line 34 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1EefRAim46YZiZlfv:-Mn1EefS27F0pffpJ75p:b-938lea) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_outbound_foreign_key.go#L34)):*
> ```Go
> 				}
> 			})),
> 		to(scpb.Status_ABSENT,
> ```

ditto comment from default expr
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_relation_depended_on_by.go, line 32 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1EfOICjcR4Mm8NeCV:-Mn1EfOICjcR4Mm8NeCW:b-938lea) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_relation_depended_on_by.go#L32)):*
> ```Go
> 				}
> 			})),
> 		to(scpb.Status_ABSENT,
> ```

ditto comment from default expr
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_sequence_owned_by.go, line 31 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1EgGrEx1cwO45ezO_:-Mn1EgGrEx1cwO45ezOa:b-938lea) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_sequence_owned_by.go#L31)):*
> ```Go
> 				}
> 			})),
> 		to(scpb.Status_ABSENT,
> ```

ditto comment from default expr
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_type_reference.go, line 32 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1EhRp4eVmk2PAU4t0:-Mn1EhRp4eVmk2PAU4t1:b-938lea) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_type_reference.go#L32)):*
> ```Go
> 				}
> 			})),
> 		to(scpb.Status_ABSENT,
> ```

ditto comment from default expr
___
*[pkg/sql/schemachanger/scplan/testdata/alter_table, line 590 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1Ew4i6cJb-wo2SDLI:-Mn1Ew4i6cJb-wo2SDLJ:bfmqiqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/testdata/alter_table#L590)):*
> ```
>     IndexID: 1
>     TableID: 53
> Stage 6
> ```

It seems wrong to me that this phase does not have the `non-revertible` tag. Shouldn't that have happened when we made the non-revertible state infectious?
___
*[pkg/sql/schemachanger/scplan/testdata/drop_database, line 53 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn10-uXBlAl5LPic2rc:-Mn1F9zL1ehHyBpnEUIm:b-o7nydj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/testdata/drop_database#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why is this non-revertible? 
</blockquote></details>

question still stangs. 


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/schemachanger/scplan/opgen/op_gen.go, line 89 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1FIQC9VLW-3So9y8N:-Mn1R8LgEPkHiQ9srLIv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/op_gen.go#L89)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 			// Revertibility is infectious and will impact
> 			// all edges that follow.
> 			if len(edgesToAdd) > 0 && !edgesToAdd[len(edgesToAdd)-1].revertible {
> 				op.revertible = false
> 			}
> ```
I thought about this some more and I no longer think it works. We don't currently add nodes that correspond to already completed stages. I think we should actually change this logic in a different place:

```go
--- a/pkg/sql/schemachanger/scplan/opgen/target.go
+++ b/pkg/sql/schemachanger/scplan/opgen/target.go
@@ -54,7 +54,7 @@ func populateSpecs(status scpb.Status, specs []transitionSpec) {
                        continue
                }
                specs[i].from = specs[i-1].to
-
+               specs[i].revertible = specs[i].revertible && specs[i-1].revertible
                prevMinPhase := specs[i-1].minPhase
                switch curMinPhase := specs[i].minPhase; curMinPhase {
                case 0:
```

</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_default_expression.go, line 39 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1ETLtEZRXIjw8XZZT:-Mn1R9AiEKOjTegzrDvB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_default_expression.go#L39)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 		to(scpb.Status_DROPPED,
> 			minPhase(scop.PreCommitPhase),
> 			revertible(false),
> 			emit(func(this *scpb.DefaultExpression) scop.Op {
> 				return &scop.RemoveColumnDefaultExpression{
> 					TableID:  this.TableID,
> 					ColumnID: this.ColumnID,
> 				}
> 			}),
> 			emit(func(this *scpb.DefaultExpression) scop.Op {
> 				return &scop.UpdateRelationDeps{
> 					TableID: this.TableID,
> 				}
> 			})),
> 		to(scpb.Status_ABSENT,
> 			minPhase(scop.PreCommitPhase),
> 			revertible(false)),
> ```
I'm not sure I understand the addition of this new phase here. I don't know what `Dropped` means here. 
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_inbound_foreign_key.go, line 34 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1EbYN5Uwmm7cvGlwp:-Mn1R9dv0FueLjnOOtwi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_inbound_foreign_key.go#L34)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

ditto comment from default expr
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_outbound_foreign_key.go, line 34 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1EefRAim46YZiZlfv:-Mn1RA345PBbeOWQbFwu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_outbound_foreign_key.go#L34)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

ditto comment from default expr
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_relation_depended_on_by.go, line 32 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1EfOICjcR4Mm8NeCV:-Mn1RAXB1BZ93ul2aTQR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_relation_depended_on_by.go#L32)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

ditto comment from default expr
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_sequence_owned_by.go, line 31 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1EgGrEx1cwO45ezO_:-Mn1RBAf5cgs-DnbVIXs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_sequence_owned_by.go#L31)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

ditto comment from default expr
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/opgen/opgen_type_reference.go, line 32 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1EhRp4eVmk2PAU4t0:-Mn1RBf_8J8bkHypt654:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/opgen/opgen_type_reference.go#L32)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

ditto comment from default expr
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/testdata/alter_table, line 31 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1-eIS13xvSxd_jQhC:-Mn1RCgk48O86kCcT1t7:b-47qbgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/testdata/alter_table#L31)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> Stage 2 (non-revertible)
>   *scop.MakeAddedColumnDeleteOnly
> ```
this isn't right, for reasons noted in the `opgen_column.go`.

On a more deep level, if we ever are force to have a non-revertible phase followed by a revertible phase, we have a bug. The `Plan` function should return an assertion failure error rather than letting this happen. 
</blockquote></details>

Ugh, I realized something stupid. Somehow the bad change was in my later commit, propagated it there and its fixed now.
___
*[pkg/sql/schemachanger/scplan/testdata/alter_table, line 590 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1Ew4i6cJb-wo2SDLI:-Mn1RIK41mPAKtCYq5E0:b8uyz7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6cbfd2e18a1bb96c4b91b42a8d417c46d81fe/pkg/sql/schemachanger/scplan/testdata/alter_table#L590)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It seems wrong to me that this phase does not have the `non-revertible` tag. Shouldn't that have happened when we made the non-revertible state infectious?
</blockquote></details>

Yes, also added an assertion to validate this when building stages.
___
*[pkg/sql/schemachanger/scplan/testdata/drop_database, line 53 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn10-uXBlAl5LPic2rc:-Mn1O48W5RvWxVAo2epv:b-o01uba) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/testdata/drop_database#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

question still stangs. 
</blockquote></details>

This was fixed with the reverability propagation fix.
___
*[pkg/sql/schemachanger/scplan/testdata/drop_database, line 86 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-Mn1039i6t8H4su4uBT0:-Mn1R4H444CeAPxCZwDd:b-yihktn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3485fbb57df9e5267d8188b86b786a06964e9b45/pkg/sql/schemachanger/scplan/testdata/drop_database#L86)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 
>   *scop.RemoveTypeBackRef
>     DescID: 64
>     TypeID: 62
>   *scop.RemoveTypeBackRef
>     DescID: 64
>     TypeID: 63
> ```
I don't think this should happen before the relevant descriptor is actually marked as dropped. 
</blockquote></details>

Looks fixed now.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I think it needs a regen of the rules. When CI is happy, I'm happy. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68616#-:-Mn1XKmL7HEmjzIvJl_m:bc35xei)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)



<!-- Sent from Reviewable.io -->
